### PR TITLE
Initial limited support for direct connections

### LINKED ISSUE #96: Expand Connection API to support direct connections

### PR DESCRIPTION
**Pull request by rhauch**
## Summary of Changes

Resolves #96

This adds the _initial but limited_ support for a new type of connection resources, called `DIRECT` connections, that represent a local or remote Kafka cluster and/or Schema Registry. The extension might create a direct connection to talk to a locally running CP installation, or a remote Apache Kafka cluster with a remote Schema Registry, or even a CCloud cluster with Schema Registry. The user may have to provide credentials to successfully authenticate and connect to each remote service/cluster.

The direction connections are not necessarily the constructs we want to expose. It is likely that near-term versions of the VS Code extension will allow users to connect to local or remote Kafka and SR clusters, but it may also expose to users different kinds of connections that are all represented in the sidecar as direct connections. For example, maybe the VS Code extension allows users to connect to “Confluent Platform”, “Apache Kafka”, and “WarpStream” (in addition to CCloud and local AK+SR), and these might all be customized facades on top of direct connections.

This is the first step in full support for direct connections, and as such only limited functionality is available in this PR:
* CRUDL connection resources with `type=DIRECT`, with validation of fields for each type of connection.
* A direct connection contains zero or one `kafka_cluster` object and zero or one `schema_registry` objects, allowing it to connect to a (remote or local) Kafka cluster and/or a (remote or local) Schema Registry.
* GraphQL support for the Kafka cluster and SR referenced by direct connections.

Once this PR is merged, the following will handled as followups to complete the support for direct connections:
- [ ] Support using Direct connections (w/ no auth) in Kafka REST API and Message Consumer API endpoints 
- [ ] Support using Direct connections (w/ no auth) in SR API endpoints.
- [ ] Update direct connection status by showing authn success against Kafka cluster and/or SR. This may mean we evolve the current `status` section of the connection resources to better work with non-CCloud connections, where connectivity to Kafka cluster and SR are independent. We'll need separate sections for each kind of cluster configured.
- [ ] Add basic credentials to direct connections via the `kafka_cluster` and `schema_registry` definitions, and use this to connect to downstream resources (e.g., Kafka REST API, SR API endpoints, and Message Consumer API). All secrets in credentials will be write-only: the Connection REST API endpoints must always mask secrets. 
- [ ] Add mTLS support to direct connections (similar to basic creds)
- [ ] Add OAuth credentials support to direct connections (similar to basic creds)
- [ ] Add Kerberos (SSL/GSSAPI) credentials support to direct connections (similar to basic creds)

After that, we will have more improvements to better support Confluent Platform: 
- [ ] Implement the existing `PLATFORM` connection type (or rename to `MDS`) that connects to an MDS URL using basic credentials and uses the metadata API to dynamically discover the available Kafka and SR clusters that were registered with MDS. Once connected, the GraphQL response will include these clusters. Like with direct connections, these connections should provide configuration property overrides for the the Kafka and SR clients.

Other improvements may include:
- [ ] Local connections can define SR URL via same `schema_registry` field as direct, superseding the existing `local_config`, which will still be supported for a while so that the extension can migrate. A local connection is invalid if it has both `local_config` and `schema_registry` objects.
- [ ] Potentially add custom client properties to `kafka_cluster` and `schema_registry`, allowing users to override some client connection properties. 
- [ ] Allow direct connections to connect to Flink cluster and Kafka Connect cluster.

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
    - [x] Updated existing
    - [x] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [ ] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rohitsanj) [inline src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:261]: Thanks for adding in the support to "automatically" go from CCloud bootstrap servers -> Kafka REST URI, it's nice that we can construct it with some RegEx and string manipulation. 

User gives us CCloud bootstrap servers and a resource API key -- sidecan then has optionality on whether to talk to the cluster directly over native protocol or use the Kafka REST API. Sweet. 

Could you also shine some light on your rationale for adding this support?

**Review by rohitsanj — CHANGES_REQUESTED** [review]
This is a great start, thank you @rhauch! I have some minor comments/questions and the rest are minor suggestions for fixes.

File: src/main/java/io/confluent/idesidecar/restapi/models/graph/DirectKafkaCluster.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/graph/DirectKafkaCluster.java:21]: Good catch.

File: src/test/java/io/confluent/idesidecar/restapi/resources/ConnectionsResourceTest.java
Comment (rhauch) [reply src/test/java/io/confluent/idesidecar/restapi/resources/ConnectionsResourceTest.java:749]: Yeah, that's what's kinda weird about the `LocalConfig` object and it's behavior. Right now, the `schema-registry-uri` can be:
* Null if the user wants to use the SR at the default `http://localhost:8082`
* Empty (zero-length string) if the user does not want to use SR
* A valid URL if the user wants to use the SR at any port

Here, this test considers it invalid if the URI contains only one or more whitespace characters. Essentially, this is not a valid URL.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:261]: The primary reason is so that the user can create a direct connection to a CCloud Kafka or SR cluster, and we're able to infer the cluster's `id` value without connecting to it. And if the CCloud cluster is a Kafka cluster, then we can also infer the Kafka REST URL, for exposure in the GraphQL responses.

But then in the sidecar we can potentially optimize the CCloud scenarios by more easily getting the various ways to connect (e.g., CCloud APIs, Kafka REST URL, etc.), without necessarily having to extract information out of the plain `KafkaCluster` object, as that will already have been done.

File: src/main/java/io/confluent/idesidecar/restapi/models/graph/DirectKafkaCluster.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/graph/DirectKafkaCluster.java:34]: I'll take your other suggestion, and remove `kafkaRestHostName`.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:229]: Same comment as above. I think it would be great if these constraints are part of the OpenAPI spec.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (flippingbits) [inline src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:468]: I got used to using names that start with `is` for methods that return a boolean value. What about renaming this method to something like `addNotAllowedError` or similar?

**Review by flippingbits — COMMENTED** [review]
Great work, @rhauch! I have only a few minor comments/suggestions.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:28]: How do we do that? The `minLength` attribute is added to the OpenAPI spec (perhaps based of `@NotEmpty`), but I can't find a way to add the min or max length to the generated OpenAPI spec.

Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:28]: Ah, I found it: `@Schema` has lots of parameters.

Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:28]: So if we're not going to use the `@Max` annotation, should we remove them?

Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:28]: > So if we're not going to use the @Max and other javax.validation.constaint annotations, should we remove them?

No, we should still keep them. The `@Schema` (coming from `org.eclipse.microprofile.openapi.annotations.media.Schema`) only provides a way to embed documentation in code and (I believe) does not automatically offer any validation logic.

Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:28]: Ah, `javax.validation.constraint.Size(...)` works.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rohitsanj) [inline src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java:21]: I just realized this, but we should be using the Jakarta flavor APIs here instead of Javax. The namespace is compatible, so just replacing `javax` -> `jakarta` should work.

Here's why (courtesy to our internal knowledge base):
The primary reason to transition to Jakarta APIs from javax is due to the transfer of Java EE to the Eclipse Foundation, where it became Jakarta EE.  The older javax packages within Java EE have been deprecated.  Jakarta EE uses the jakarta namespace, which is incompatible with the legacy javax namespace.  Mixing the two can cause issues.  Several projects and frameworks, including Jetty 12, Spring 6, Spring Boot 3, Apache Tomcat 11, and ActiveMQ 6, have already adopted Jakarta.  Additionally, upgrading to Jakarta allows you to keep up with the latest changes and modernize your applications.  In the context of Apache Kafka, upgrading to Jakarta in version 4.0 was considered a good time to make the change as it required incompatible package changes, aligning with the major version release.

**Review by rohitsanj — CHANGES_REQUESTED** [review]

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java]: Ah, that's nice! Glad we don't have to maintain those limits in two annotations then.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java]: I renamed all of the methods to `check*(...)`, which I think is a decent combination of readable that avoids name patterns that are used for getters.

File: src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java
Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java]: Further reading (posting for myself to go read later):

- https://jakarta.ee/blogs/javax-jakartaee-namespace-ecosystem-progress/
- https://blog.ltgt.net/javax-jakarta-mess-and-gradle-solution/, and its sequel, https://blog.ltgt.net/javax-jakarta-mess-even-worse/

Comment (rhauch) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java]: Thanks for pointing that out. I've fixed this PR to use `jakarta.validation.*` instead of `javax.validation`, but we have a few other instances of `javax` imports. I'd prefer to make those changes in a separate PR.

https://github.com/jakartaee/platform/blob/main/namespace/mappings.adoc#mapping-javax-to-jakarta

Comment (rohitsanj) [reply src/main/java/io/confluent/idesidecar/restapi/models/ConnectionSpec.java]: That's fine and yeah I noticed those too. I believe javax is a transitive dependency so until those dependencies get rid of it, we'll keep running into this. I guess a custom checkstyle rule can prevent us from running into it again.

**Review by rohitsanj — APPROVED** [review]
Thanks! LGTM 🙌

**Comment by rhauch** [comment]
Waiting for final approval from @flippingbits, since they had some questions.

**Review by flippingbits — APPROVED** [review]
Thanks for the update. LGTM, @rhauch!

**Comment by rhauch** [comment]
#112 was recently merged. Locally, I did the following to verify `main` builds with this PR branch:
```
$ git pull origin main
$ git checkout direct-connect
$ git merge main
$ make clean build test mvn-package-native
```
All is good, and I did some basic testing with the native image, confirming that the Connections API works as expected for local, CCloud and the new direct connections. The local and CCloud connections should not be affected by this PR build, other than validating create and update of connection resources, which have good coverage with new tests.

Based upon this, I'm merging this PR.

**Pull request merged**
